File: src-tauri/src/models/awake.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum AwakeBehavior {
    PreventSystemSleep,
    KeepDisplayAwake,
}

impl AwakeBehavior {
    pub fn display_name(&self) -> &'static str {
        match self {
            AwakeBehavior::PreventSystemSleep => "Keep Computer Awake",
            AwakeBehavior::KeepDisplayAwake => "Keep Computer + Display Awake",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum PowerCondition {
    #[default]
    Always,
    AcPowerOnly,
}

impl PowerCondition {
    pub fn display_name(&self) -> &'static str {
        match self {
            Self::Always => "Always",
            Self::AcPowerOnly => "Only while plugged in",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum PowerSourceType {
    Ac,
    Battery,
    #[default]
    Unknown,
}

impl PowerSourceType {
    pub fn is_ac(&self) -> bool {
        matches!(self, Self::Ac)
    }

    pub fn display_name(&self) -> &'static str {
        match self {
            Self::Ac => "Plugged In (AC)",
            Self::Battery => "Battery Power",
            Self::Unknown => "Unknown Power Source",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum AwakeRuleStatus {
    Active,
    WaitingApplication,
    WaitingAgent,
    WaitingProcess,
    WaitingPower,
    InvalidApplication,
    Disabled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum AwakeAgentId {
    Codex,
    Claude,
    Antigravity,
    #[serde(rename = "opencode")]
    OpenCode,
}

impl AwakeAgentId {
    pub const ALL: [Self; 4] = [Self::Codex, Self::Claude, Self::Antigravity, Self::OpenCode];

    pub fn display_name(self) -> &'static str {
        match self {
            Self::Codex => "Codex",
            Self::Claude => "Claude Code",
            Self::Antigravity => "Antigravity",
            Self::OpenCode => "OpenCode / OMP",
        }
    }

    /// The adapter id is shared with Agent Activity so a typed Keep Awake rule
    /// cannot invent its own process-signature table.
    pub const fn adapter_id(self) -> &'static str {
        match self {
            Self::Codex => "codex",
            Self::Claude => "claude",
            Self::Antigravity => "antigravity",
            Self::OpenCode => "opencode",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
pub struct ApplicationIdentity {
    pub display_name: String,
    pub executable_name: String,
    pub path: String,
}

impl ApplicationIdentity {
    pub fn is_structurally_valid(&self) -> bool {
        !self.display_name.trim().is_empty()
            && !self.executable_name.trim().is_empty()
            && std::path::Path::new(&self.path).is_absolute()
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
pub struct AwakeRule {
    pub id: String,
    pub app_name: String,
    pub executable_pattern: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_process_pattern: Option<String>,
    /// Native picker identity for the primary application. When present, this
    /// is a typed rule and raw pattern fields are retained only for legacy
    /// compatibility/debugging; they are not used for matching.
    #[serde(default)]
    pub application: Option<ApplicationIdentity>,
    /// Allowlisted agent adapters. Semantics are explicit any-of (OR).
    #[serde(default)]
    pub agent_ids: Vec<AwakeAgentId>,
    pub behavior: AwakeBehavior,
    #[serde(default)]
    pub power_condition: PowerCondition,
    pub enabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
pub struct SelectedApplication {
    pub name: String,
    pub executable_pattern: String,
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
pub struct AwakeRuleEvaluation {
    pub rule_id: String,
    pub status: AwakeRuleStatus,
    pub is_process_running: bool,
    pub is_power_eligible: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
pub struct AwakeState {
    pub is_active: bool,
    pub behavior: Option<AwakeBehavior>,
    pub trigger_source: Option<String>,
    pub active_process_name: Option<String>,
    pub active_rule_id: Option<String>,
    #[serde(with = "crate::ipc_numeric::option_u64")]
    #[specta(type = Option<u64>)]
    pub manual_expires_at: Option<u64>,
    pub active_rules_count: usize,
    pub power_source: PowerSourceType,
    pub last_error: Option<String>,
    pub rule_evaluations: Vec<AwakeRuleEvaluation>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn backward_compatible_deserialization_defaults_to_always() {
        let legacy_json = r#"{
            "id": "legacy.rule",
            "app_name": "Legacy App",
            "executable_pattern": "legacy_bin",
            "behavior": "prevent_system_sleep",
            "enabled": true
        }"#;

        let rule: AwakeRule = serde_json::from_str(legacy_json).unwrap();
        assert_eq!(rule.power_condition, PowerCondition::Always);
        assert_eq!(rule.requires_process_pattern, None);
        assert_eq!(rule.application, None);
        assert!(rule.agent_ids.is_empty());
    }

    #[test]
    fn typed_rule_round_trips_application_identity_and_known_agents() {
        let rule = AwakeRule {
            id: "rule.warp-agents".into(),
            app_name: "Warp".into(),
            executable_pattern: "Warp".into(),
            requires_process_pattern: None,
            application: Some(ApplicationIdentity {
                display_name: "Warp".into(),
                executable_name: "stable".into(),
                path: "/Applications/Warp.app".into(),
            }),
            agent_ids: vec![AwakeAgentId::Codex, AwakeAgentId::OpenCode],
            behavior: AwakeBehavior::PreventSystemSleep,
            power_condition: PowerCondition::AcPowerOnly,
            enabled: false,
        };

        let encoded = serde_json::to_string(&rule).unwrap();
        let decoded: AwakeRule = serde_json::from_str(&encoded).unwrap();
        assert_eq!(decoded, rule);
        assert!(decoded.application.unwrap().is_structurally_valid());
    }
}

File: src/lib/utils/awake.ts
import type { ApplicationIdentity, AwakeAgentId, AwakeRule } from '../models/types';

export const AWAKE_AGENT_OPTIONS: ReadonlyArray<{
  id: AwakeAgentId;
  label: string;
  hint: string;
}> = [
  { id: 'codex', label: 'Codex', hint: 'Codex CLI activity' },
  { id: 'claude', label: 'Claude Code', hint: 'Claude Code activity' },
  { id: 'antigravity', label: 'Antigravity', hint: 'agy / antigravity activity' },
  { id: 'opencode', label: 'OpenCode / OMP', hint: 'opencode / omp activity' },
];

const agentLabels = new Map<AwakeAgentId, string>(
  AWAKE_AGENT_OPTIONS.map((agent) => [agent.id, agent.label])
);

export function awakeAgentLabel(agentId: AwakeAgentId): string {
  return agentLabels.get(agentId) ?? agentId;
}

export function selectedAwakeAgents(rule: AwakeRule): AwakeAgentId[] {
  return [...(rule.agent_ids ?? [])];
}

export function toggleAwakeAgent(
  selected: readonly AwakeAgentId[],
  agentId: AwakeAgentId
): AwakeAgentId[] {
  return selected.includes(agentId)
    ? selected.filter((candidate) => candidate !== agentId)
    : [...selected, agentId];
}

function applicationLabel(application: ApplicationIdentity | null | undefined): string {
  return application?.display_name?.trim() || 'this application';
}

function agentClause(agentIds: readonly AwakeAgentId[]): string {
  if (agentIds.length === 0) return '';
  const labels = agentIds.map(awakeAgentLabel);
  const joined = labels.length === 1
    ? labels[0]
    : labels.length === 2
      ? `${labels[0]} or ${labels[1]}`
      : `${labels.slice(0, -1).join(', ')}, or ${labels.at(-1)}`;
  return ` and any of ${joined} is active`;
}

function powerClause(powerCondition: AwakeRule['power_condition']): string {
  return powerCondition === 'ac_power_only'
    ? 'only while plugged in'
    : 'on AC or battery power';
}

function behaviorClause(behavior: AwakeRule['behavior']): string {
  return behavior === 'keep_display_awake'
    ? 'The display stays awake.'
    : 'The display may sleep.';
}

/**
 * Creates the user-facing contract for typed rules. It deliberately ignores
 * raw legacy matcher fields so the basic flow never leaks pipe syntax or
 * command-line matching details.
 */
export function awakeRuleSummary(rule: AwakeRule): string {
  if (!rule.application) {
    return `Custom process rule: ${rule.app_name || 'unnamed rule'}.`;
  }
  return `Keep this computer awake when ${applicationLabel(rule.application)} is open${agentClause(
    selectedAwakeAgents(rule)
  )}, ${powerClause(rule.power_condition)}. ${behaviorClause(rule.behavior)}`;
}

export function typedAwakeRule(rule: AwakeRule): boolean {
  return Boolean(rule.application);
}
